lib/kube/proxy/forwarder.go

```diff
@@ -1489,22 +1522,21 @@ func (f *Forwarder) newClusterSessionLocal(ctx authContext) (*clusterSession, er
 	return sess, nil
 }
 
-func (f *Forwarder) newClusterSessionDirect(ctx authContext, kubeService types.Server) (*clusterSession, error) {
-	f.log.WithFields(log.Fields{
-		"kubernetes_service.name": kubeService.GetName(),
-		"kubernetes_service.addr": kubeService.GetAddr(),
-	}).Debugf("Kubernetes session for %v forwarded to remote kubernetes_service instance.", ctx)
+func (f *Forwarder) newClusterSessionDirect(ctx authContext, endpoints []endpoint) (*clusterSession, error) {
+	if len(endpoints) == 0 {
+		return nil, trace.BadParameter("no kube cluster endpoints provided")
+	}
+
+	f.log.WithField("kube_service.endpoints", endpoints).Debugf("Kubernetes session for %v forwarded to remote kubernetes_service instance.", ctx)
+
 	sess := &clusterSession{
 		parent:      f,
 		authContext: ctx,
 		// This session talks to a kubernetes_service, which should handle
 		// audit logging. Avoid duplicate logging.
 		noAuditEvents: true,
 	}
-	// Set both addr and serverID, in case this is a kubernetes_service
-	// connected over a tunnel.
-	sess.authContext.teleportCluster.targetAddr = kubeService.GetAddr()
-	sess.authContext.teleportCluster.serverID = fmt.Sprintf("%s.%s", kubeService.GetName(), ctx.teleportCluster.name)
+	sess.authContext.teleportClusterEndpoints = endpoints
```

Comment:
Do we need to synchronize access to `sess.authContext` ?

Reply:
I could be wrong but I don't see any race conditions as is. It looks to me like the proxy-forwarding logic is all asynchronous since it is all run on each request.